### ISSUE TITLE
Implement custom resource lifecycle methods in Python `Provider` class

### ISSUE THREAD

**Issue by ringods** [pulumi/pulumi#13977]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

A customer is blocked on:

* #1064 
* #7960 

Fixing the Python pickling doesn't seem to be an easy fix. As an alternative, I promoted moving their Python dynamic provider classes to a true Python based Pulumi package would be an option.

We have a [Pulumi Component Boilerplate (Python)](https://github.com/pulumi/pulumi-component-provider-py-boilerplate) which makes use of the following classes from the Pulumi Python SDK:

* [ProviderServicer](https://github.com/pulumi/pulumi/blob/master/sdk/python/lib/pulumi/provider/server.py)
* [Provider](https://github.com/pulumi/pulumi/blob/master/sdk/python/lib/pulumi/provider/provider.py)

These classes only have the methods implemented to support Python based components (`Construct`).

The request is to complete the implementation of the remaining methods to also support custom resources (`Create`, `Update`, `Delete`, ...). Having a complete implementation of these classes opens a path to unblock the customer.

### Affected area/feature

* Pulumi Python SDK

**Comment by arubino** [comment]
Hello, is there any update on this issue?

**Comment by Frassle** [comment]
Yes, I've been working on it this week. The branch so far is at https://github.com/pulumi/pulumi/pull/19080.

**Comment by arubino** [comment]
That's great news, thanks! Do you think we would have something available end of april?

**Comment by Frassle** [comment]
> Do you think we would have something available end of April?

Yes

**Comment by angelopoerio** [comment]
Thanks for the update. Once the PR is merged a new version of the pulumi library will be released as well and ready to be used? Will we need to update the pulumi runtime as well?

**Comment by Frassle** [comment]
> Once the PR is merged a new version of the pulumi library will be released as well and ready to be used? Will we need to update the pulumi runtime as well?

This doesn't change anything in the pulumi CLI. It's just a python library update.

**Comment by arubino** [comment]
Hi @Frassle, are there any news on this? 🙂 thanks!

**Comment by Frassle** [comment]
Still working on it, we've got meetings about it later today. Still hopeful we can get this done in the next week or so.
